rust/osm2lanes/src/transform/tags_to_lanes/mod.rs

```diff
@@ -565,59 +568,191 @@ pub fn tags_to_lanes(
     Ok(road_from_tags)
 }
 
-fn driving_lane_directions(tags: &Tags, _locale: &Locale, oneway: Oneway) -> (usize, usize) {
-    let both_ways = if let Some(n) = tags
-        .get("lanes:both_ways")
-        .and_then(|num| num.parse::<usize>().ok())
-    {
-        n
-    } else {
-        0
-    };
-    let num_driving_fwd = if let Some(n) = tags
-        .get("lanes:forward")
-        .and_then(|num| num.parse::<usize>().ok())
-    {
-        n
-    } else if let Some(n) = tags.get("lanes").and_then(|num| num.parse::<usize>().ok()) {
-        let half = if oneway.into() {
-            n
-        } else {
-            // usize division rounded up
-            (n + 1) / 2
-        };
-        half - both_ways
-    } else if tags.is("lanes:bus", "2") {
-        2
+impl Tags {
+    /// Gets the value for the given key and parses it into T.
+    fn get_and_parse<T: FromStr>(&self, key: &str) -> Option<T> {
+        self.get(key).and_then(|val| val.parse::<T>().ok())
+    }
+
+    /// Gets the value for the given key and parses it into T. A RoadMsg::Unsupported is added if
+    /// parsing fails.
+    fn get_parsed<T: FromStr>(&self, key: &str, warnings: &mut RoadWarnings) -> Option<T> {
+        self.get(key).and_then(|val| match val.parse::<T>() {
+            Ok(n) => Some(n),
+            Err(_) => {
+                warnings.push(RoadMsg::unsupported_tag(key.to_owned(), val));
+                None
+            }
+        })
+    }
+}
+
+/// Calculates the number of vehicle travel lanes in each direction (forward, both ways, backward)
+/// according to the `lanes` schema (which excludes parking lanes, bike lanes, etc.).
+/// See https://wiki.openstreetmap.org/wiki/Key:lanes.
+///
+/// Validates `lanes[:{forward,both_ways,backward}]=*` to determine the precise answer.
+/// Uses `lanes:{bus,psv}[:{forward,backward}]=*`, `busway[:{left,both,right}]=*` and
+/// `centre_turn_lane=yes` to make assumptions in the absence of precise tagging.
+fn driving_lane_directions(
+    tags: &Tags,
+    locale: &Locale,
+    oneway: Oneway,
+    warnings: &mut RoadWarnings,
+) -> (usize, usize, usize) {
```

Comment:
can we test this in isolation, it is quite complex now :)

Reply:
I agree! (if this function ends up keeping all of its responsibilities.)

Does that mean a set of tests in tests.yaml, or rust tests?

Reply:
Rust tests I'd imagine, since the input and output are quite specialized here. The unit tests can be in this file, since it's private